pkg/util/namespaces/user_linux.go

```diff
@@ -0,0 +1,59 @@
+// Copyright (c) 2019, Sylabs Inc. All rights reserved.
+// This software is licensed under a 3-clause BSD license. Please consult the
+// LICENSE.md file distributed with the sources of this project regarding your
+// rights to use or distribute this software.
+
+package namespaces
+
+import (
+	"bufio"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"strconv"
+	"strings"
+)
+
+// ParentUserNamespace checks if a process is already running in a
+// parent user namespace, also returns if the process has permissions
+// to use setgroups.
+func ParentUserNamespace(pid int) (bool, bool) {
+	// default values returned in case of error
+	parent := false
+	setgroups := false
+
+	// can fail if the kernel doesn't support user namespace
+	r, err := os.Open(fmt.Sprintf("/proc/%d/uid_map", pid))
+	if err != nil {
+		return parent, setgroups
+	}
+	defer r.Close()
+
+	scanner := bufio.NewScanner(r)
+	for scanner.Scan() {
+		fields := strings.Fields(scanner.Text())
+
+		// trust values returned by procfs
+		size, _ := strconv.ParseUint(fields[2], 10, 32)
+
+		// a size of 4294967295 means the process is running
+		// in the host user namespace
+		if uint32(size) == ^uint32(0) {
+			return parent, setgroups
+		}
+
+		// we are in a parent user namespace
+		parent = true
+
+		// should not fail if open call passed
```

Comment:
shouldn't this be a different function? Like CanSetGroups or something like that? I understand that the two things are related, but the new code shows there are cases where you care about the first value and not the second.

Reply:
No because it could be used wrongly because `/proc/pid/setgroups` is set to `allow` by default for process doesn't living inside a user namespace, so we only check read setgroups value once we have confirmation that we are indeed in a user namespace

Reply:
good, thanks